src/petals/server/backend.py

```diff
@@ -59,24 +60,31 @@ def inference_step(self, cache_metadata: torch.IntTensor, *inputs: torch.Tensor)
             ), "expected hidden states to be 3-dimensional: [batch_size, seq_len, hid_size]"
 
             with self.memory_cache.use_cache(attention_cache_handle) as cache:
-                assert isinstance(self.module, BloomBlock) and cache.shape[0] == 2 and cache.ndim == 5
+                batch_size = cache.shape[1]
+                max_length = cache.numel() // (2 * batch_size * head_dim * num_heads)
+                assert isinstance(self.module, WrappedBloomBlock) and cache.shape[0] == 2 and cache.ndim == 3
                 if not is_dummy(hypo_ids):
                     assert hypo_ids.shape[0] == cache.shape[1]
                     cache[:, :] = cache[:, hypo_ids]  # in-place reorder cache by hypo ids
-                layer_past = past_k, past_v = cache[0, :, :prefix_length], cache[1, :, :prefix_length]
-                logger.debug(f"Metadata: {cache_metadata}, past_k.shape={past_k.shape}, past_v.shape={past_v.shape}")
-                hidden_states, (new_k, new_v) = self.module.forward(
-                    hidden_states, layer_past=layer_past, use_cache=True
-                )
+                key_cache = cache[0].view(batch_size, num_heads, head_dim, max_length)
+                value_cache = cache[1].view(batch_size, num_heads, max_length, head_dim)
 
-                # todo remove these asserts once we pass all tests
-                new_length = new_v.shape[1]
+                key_past = key_cache.flatten(0, 1)[:, :, :prefix_length]  # [batch * num_heads, head_dim, kv_length]
+                value_past = value_cache.flatten(0, 1)[:, :prefix_length, :]  # [batch * num_heads, kv_length, head_dim]
```

Comment:
Can't you just directly reshape the past tensors to these shapes like you've done in `src/petals/server/handler.py`?

Reply:
Nope, we cannot
- hypo_ids need shape [2, batch_size, ...]
- training needs key [batch_size * heads, ..., length] and value [..., length, :], making them non-concat-able
- handler needs them to be concat-able in a single tensor